packages/superset-ui-core/src/callApi/parseResponse.js

```diff
@@ -1,26 +1,29 @@
-export default function parseResponse(apiPromise) {
-  return apiPromise.then(apiResponse =>
+const PARSERS = {
+  json: response =>
     // first try to parse as json, and fall back to text (e.g., in the case of HTML stacktrace)
     // cannot fall back to .text() without cloning the response because body is single-use
-    apiResponse
+    response
       .clone()
       .json()
-      .catch(() => /* jsonParseError */ apiResponse.text().then(textPayload => ({ textPayload })))
-      .then(maybeJson => ({
-        json: maybeJson.textPayload ? undefined : maybeJson,
-        response: apiResponse,
-        text: maybeJson.textPayload,
-      }))
-      .then(({ response, json, text }) => {
-        if (!response.ok) {
-          return Promise.reject({
-            error: response.error || (json && json.error) || text || 'An error occurred',
-            status: response.status,
-            statusText: response.statusText,
-          });
-        }
+      .then(json => ({ json, response }))
+      .catch(() => /* jsonParseError */ response.text().then(text => ({ response, text }))),
```

Comment:
Does this mean the developer always have to check if `output.json` exists or not? If `output.json` does not exist, they should use to `output.text` instead. If they forget to check. This can cause bugs. I wonder if there is a way to make it more explicit. 

Also, generally, does one expect this Promise to reject when fail to parse or resolve text of the supposed-to-be json? IMO, I would expect it to fail and give me the text in the error message for debugging.

Reply:
Yeah now that we expose the parse method and are explicit about `text`/`json` we should probably _not_ clone and fallback to `text` gracefully, and instead it should probably throw.

Reply:
this broke a lot of tests, which were falling back to `.text()` unknowingly because of my mock text response 😜